cpu/efm32/periph/i2c.c

```diff
@@ -122,128 +126,102 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
 
 int i2c_acquire(i2c_t dev)
 {
+    /* acquire lock */
     mutex_lock(&i2c_lock[dev]);
 
+    /* power peripheral */
+    CMU_ClockEnable(i2c_config[dev].cmu, true);
+
     return 0;
 }
 
 int i2c_release(i2c_t dev)
 {
+    /* disable peripheral */
+    CMU_ClockEnable(i2c_config[dev].cmu, false);
+
+    /* release lock */
     mutex_unlock(&i2c_lock[dev]);
 
     return 0;
 }
 
-int i2c_read_byte(i2c_t dev, uint8_t address, void *data)
+int i2c_read_byte(i2c_t dev, uint16_t address, void *data, uint8_t flags)
 {
-    return i2c_read_bytes(dev, address, data, 1);
+    return i2c_read_bytes(dev, address, data, 1, flags);
 }
 
-int i2c_read_bytes(i2c_t dev, uint8_t address, void *data, int length)
+int i2c_read_bytes(i2c_t dev, uint16_t address, void *data, size_t length, uint8_t flags)
 {
     I2C_TransferSeq_TypeDef transfer;
 
     transfer.addr = (address << 1);
-    transfer.flags = I2C_FLAG_READ;
+    transfer.flags = I2C_FLAG_READ | (flags & I2C_ADDR10 ? I2C_FLAG_10BIT_ADDR : 0);
     transfer.buf[0].data = (uint8_t *) data;
     transfer.buf[0].len = length;
 
     /* start a transfer */
-    _transfer(dev, &transfer);
-
-    if (i2c_progress[dev] != i2cTransferDone) {
-        return -2;
-    }
-
-    return length;
+    return _transfer(dev, &transfer);
 }
 
-int i2c_read_reg(i2c_t dev, uint8_t address, uint8_t reg, void *data)
+int i2c_read_reg(i2c_t dev, uint16_t address, uint16_t reg, void *data, uint8_t flags)
 {
-    return i2c_read_regs(dev, address, reg, data, 1);
+    return i2c_read_regs(dev, address, reg, data, 1, flags);
 }
 
-int i2c_read_regs(i2c_t dev, uint8_t address, uint8_t reg,
-                  void *data, int length)
+int i2c_read_regs(i2c_t dev, uint16_t address, uint16_t reg,
+                  void *data, size_t length, uint8_t flags)
 {
     I2C_TransferSeq_TypeDef transfer;
 
     transfer.addr = (address << 1);
-    transfer.flags = I2C_FLAG_WRITE_READ;
-    transfer.buf[0].data = &reg;
-    transfer.buf[0].len = 1;
+    transfer.flags = I2C_FLAG_WRITE_READ | (flags & I2C_ADDR10 ? I2C_FLAG_10BIT_ADDR : 0);
+    transfer.buf[0].data = (uint8_t *) &reg;
+    transfer.buf[0].len = flags & I2C_REG16 ? 2 : 1;
     transfer.buf[1].data = (uint8_t *) data;
     transfer.buf[1].len = length;
 
     /* start a transfer */
-    _transfer(dev, &transfer);
-
-    if (i2c_progress[dev] != i2cTransferDone) {
-        return -2;
-    }
-
-    return length;
+    return _transfer(dev, &transfer);
 }
 
-int i2c_write_byte(i2c_t dev, uint8_t address, uint8_t data)
+int i2c_write_byte(i2c_t dev, uint16_t address, uint8_t data, uint8_t flags)
```

Comment:
You don't need this function, it's already provided by the common i2c driver (in drivers/periph_common/)